File: test/fingerprint.test.ts
import fs from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import { afterEach, describe, expect, it } from "vitest";
import { sourceFingerprint } from "../src/fingerprint.js";
import { git } from "../src/git.js";

const temporaryDirectories: string[] = [];

afterEach(async () => {
  await Promise.all(
    temporaryDirectories
      .splice(0)
      .map((directory) => fs.rm(directory, { recursive: true, force: true })),
  );
});

describe("candidate source fingerprint", () => {
  it("supports repositories before ship-my-flutter is initialized", async () => {
    const root = await fs.mkdtemp(path.join(os.tmpdir(), "smf-fingerprint-"));
    temporaryDirectories.push(root);
    await git(root, ["init", "-b", "main"]);
    await fs.writeFile(path.join(root, "lib.dart"), "void main() {}\n");
    await git(root, ["add", "."]);
    await expect(sourceFingerprint(root)).resolves.toMatch(/^[a-f0-9]{64}$/u);
  });

  it("ignores human-editable notes and candidate receipts", async () => {
    const root = await fs.mkdtemp(path.join(os.tmpdir(), "smf-fingerprint-"));
    temporaryDirectories.push(root);
    await git(root, ["init", "-b", "main"]);
    await fs.mkdir(path.join(root, ".ship-my-flutter", "candidates"), {
      recursive: true,
    });
    await fs.writeFile(path.join(root, "lib.dart"), "void main() {}\n");
    await fs.writeFile(
      path.join(root, ".ship-my-flutter", "store-release-notes.json"),
      "{}\n",
    );
    await fs.writeFile(
      path.join(root, ".ship-my-flutter", "config.json"),
      `${JSON.stringify({
        platforms: {
          ios: {
            projectPath: ".",
            bundleId: "dev.example.app",
            buildArgs: [],
            testflight: { groups: [] },
            appStore: { mode: "upload-only", releaseType: "manual" },
          },
        },
      })}\n`,
    );
    await fs.writeFile(
      path.join(root, ".ship-my-flutter", "candidates", "ios-1.0.0.json"),
      "{}\n",
    );
    await git(root, ["add", "."]);

    const before = await sourceFingerprint(root);
    await fs.writeFile(
      path.join(root, ".ship-my-flutter", "store-release-notes.json"),
      '{"ios":{}}\n',
    );
    await fs.writeFile(
      path.join(root, ".ship-my-flutter", "candidates", "ios-1.0.0.json"),
      '{"build":"1"}\n',
    );
    await fs.writeFile(
      path.join(root, ".ship-my-flutter", "config.json"),
      `${JSON.stringify({
        platforms: {
          ios: {
            projectPath: ".",
            bundleId: "dev.example.app",
            buildArgs: [],
            testflight: { groups: ["Internal"] },
            appStore: {
              mode: "submit-for-review",
              releaseType: "automatic",
            },
          },
        },
      })}\n`,
    );
    expect(await sourceFingerprint(root)).toBe(before);

    await fs.writeFile(
      path.join(root, ".ship-my-flutter", "config.json"),
      `${JSON.stringify({
        platforms: {
          ios: {
            projectPath: ".",
            bundleId: "dev.example.app",
            buildArgs: ["--dart-define=ENV=production"],
          },
        },
      })}\n`,
    );
    expect(await sourceFingerprint(root)).not.toBe(before);

    await fs.writeFile(path.join(root, "lib.dart"), "void main() => run();\n");
    expect(await sourceFingerprint(root)).not.toBe(before);
  });
});

File: test/fingerprint_test.dart
import 'dart:io';

import 'package:path/path.dart' as p;
import 'package:ship_my_flutter/ship_my_flutter.dart';
import 'package:test/test.dart';

String configYaml({
  List<String> buildArgs = const <String>[],
  List<String> groups = const <String>[],
  String mode = 'upload-only',
  String releaseType = 'manual',
}) =>
    '''
platforms:
  ios:
    projectPath: .
    bundleId: dev.example.app
    buildArgs:
${buildArgs.map((String value) => '      - "$value"').join('\n')}
    testflight:
      groups:
${groups.map((String value) => '        - "$value"').join('\n')}
    appStore:
      mode: $mode
      releaseType: $releaseType
''';

void main() {
  group('candidate source fingerprint', () {
    test('supports repositories before initialization', () async {
      final root = await Directory.systemTemp.createTemp('smf-fingerprint-');
      addTearDown(() => root.delete(recursive: true));
      await git(root.path, const <String>['init', '-b', 'main']);
      await File(
        p.join(root.path, 'lib.dart'),
      ).writeAsString('void main() {}\n');
      await git(root.path, const <String>['add', '.']);
      expect(await sourceFingerprint(root.path), matches(r'^[a-f0-9]{64}$'));
    });

    test('ignores human-editable notes and candidate receipts', () async {
      final root = await Directory.systemTemp.createTemp('smf-fingerprint-');
      addTearDown(() => root.delete(recursive: true));
      await git(root.path, const <String>['init', '-b', 'main']);
      final paths = resolveShipPaths(root.path);
      await Directory(paths.candidates).create(recursive: true);
      await File(
        p.join(root.path, 'lib.dart'),
      ).writeAsString('void main() {}\n');
      await File(paths.storeReleaseNotes).writeAsString('{}\n');
      await File(paths.config).writeAsString(configYaml());
      final receipt = File(candidatePath(root.path, Platform.ios, '1.0.0'));
      await receipt.writeAsString('{}\n');
      await git(root.path, const <String>['add', '.']);

      final before = await sourceFingerprint(root.path);
      await File(paths.storeReleaseNotes).writeAsString('{"ios":{}}\n');
      await receipt.writeAsString('{"build":"1"}\n');
      await File(paths.config).writeAsString(
        configYaml(
          groups: const <String>['Internal'],
          mode: 'submit-for-review',
          releaseType: 'automatic',
        ),
      );
      expect(await sourceFingerprint(root.path), before);

      await File(paths.config).writeAsString(
        configYaml(buildArgs: const <String>['--dart-define=ENV=production']),
      );
      expect(await sourceFingerprint(root.path), isNot(before));

      await File(
        p.join(root.path, 'lib.dart'),
      ).writeAsString('void main() => run();\n');
      expect(await sourceFingerprint(root.path), isNot(before));
    });
  });
}
